### DISCUSSION TITLE
closed_trades and sl attribute

### DISCUSSION THREAD

**Discussion by msloco** [kernc/backtesting.py#862] [General]
It seems that under some conditions (that is when a trade is closed for hitting a sl) the very .sl property of the closed_trade item corresponding to the original closed trade is set to None.
I am trying to add info to closed trades in as far as the reason they got closed (strategy, tp, sl) - which i think is somehow useful when dealing with complex strategies, and I was trying to manipulate the closed_trades adding some attributes, but I noticed this strange behavior. 
Is there  a possible explanation ?
cheers

**Comment by kernc** [comment]
> Is there a possible explanation ?

I suspect, since `Trade.sl` and `.tp` are computed properties that depend on the underlying contingent order:
https://github.com/kernc/backtesting.py/blob/8c13d005a2c07ce6bb5d6c331a891f72948d723b/backtesting/backtesting.py#L657-L666
The properties are set to None here when the trade is closed:
https://github.com/kernc/backtesting.py/blob/8c13d005a2c07ce6bb5d6c331a891f72948d723b/backtesting/backtesting.py#L984-L988
Please confirm.

**Reply by msloco** [reply]
now, from what I see from the debugging session I setup:

it led me to see that beginning with this call trace:

`next->_process_orders->if stop_price: -> order._replace(stop_price=None)
`

at this point the order in question is contingent and has not "sl" property by itself. though the referred "parent trade" correctly has.
but, within the order._replace, the parent trade loses its stop, resulting to be set to "None". Consequently, upon closing, the trade.sl is already None - although I am not sure how the properties propagate from order to trade.

thanks for further clarification on this point
